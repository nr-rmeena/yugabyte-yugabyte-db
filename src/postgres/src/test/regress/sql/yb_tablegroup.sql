--
-- YB_TABLEGROUP Testsuite: Testing Statments for TABLEGROUP.
--
\h CREATE TABLEGROUP
\h ALTER TABLEGROUP
\h DROP TABLEGROUP

--
-- pg_catalog alterations. Validate columns of pg_yb_tablegroup and oids.
--
\d pg_yb_tablegroup
SELECT oid, relname, reltype, relnatts FROM pg_class WHERE relname IN ('pg_yb_tablegroup', 'pg_yb_tablegroup_oid_index');
SELECT oid, typname, typrelid FROM pg_type WHERE typname LIKE 'pg_yb_tablegroup';
--
-- CREATE TABLEGROUP
--

CREATE TABLEGROUP tgroup1;
CREATE TABLEGROUP tgroup2;
CREATE TABLEGROUP tgroup3;
CREATE TABLE tgroup_test1 (col1 int, col2 int) TABLEGROUP tgroup1;
CREATE TABLE tgroup_test2 (col1 int, col2 int) TABLEGROUP tgroup1;
CREATE TABLE nogroup (col1 int) NO TABLEGROUP; -- fail
SELECT grpname FROM pg_yb_tablegroup;
SELECT relname
    FROM (SELECT relname, unnest(reloptions) AS opts FROM pg_class) s
    WHERE opts LIKE '%tablegroup%';
CREATE INDEX ON tgroup_test1(col2);
CREATE TABLE tgroup_test3 (col1 int, col2 int) TABLEGROUP tgroup2;
-- Index opt out - should not show up in following SELECT
CREATE INDEX ON tgroup_test3(col1) NO TABLEGROUP;
-- Index explicitly specify tablegroup other than that of indexed table
CREATE INDEX ON tgroup_test3(col1) TABLEGROUP tgroup1;
SELECT s.relname, pg_yb_tablegroup.grpname
    FROM (SELECT relname, unnest(reloptions) AS opts FROM pg_class) s, pg_yb_tablegroup
    WHERE opts LIKE CONCAT('%tablegroup=', CAST(pg_yb_tablegroup.oid AS text), '%');
-- These should fail.
CREATE TABLEGROUP tgroup1;
CREATE TABLE tgroup_test (col1 int, col2 int) TABLEGROUP bad_tgroupname;
CREATE TABLE tgroup_optout (col1 int, col2 int) WITH (colocated=false) TABLEGROUP tgroup1;
CREATE TABLE tgroup_optout (col1 int, col2 int) WITH (colocated=true) TABLEGROUP tgroup1;
CREATE TABLE tgroup_optout (col1 int, col2 int) WITH (colocated=false) TABLEGROUP bad_tgroupname;
CREATE TEMP TABLE tgroup_temp (col1 int, col2 int) TABLEGROUP tgroup1;

-- Can use WITH to create a tablegroup
CREATE TABLE tgroup_with1 (col1 int, col2 int) WITH (tablegroup=16385);
-- Cannot use tablegroups and colocated=true/false
CREATE TABLE tgroup_with2 (col1 int, col2 int) WITH (tablegroup=16385, colocated=true);
CREATE TABLE tgroup_with2 (col1 int, col2 int) WITH (tablegroup=16385, colocated=false);
-- Cannot specify tablegroup OID and tablegroup name
CREATE TABLE tgroup_with3 (col1 int, col2 int) WITH (tablegroup=16385) TABLEGROUP tgroup1;
-- Cannot use an invalid tablegroup OID
CREATE TABLE tgroup_with4 (col1 int, col2 int) WITH (tablegroup=123);

--
-- Cannot drop dependent objects
--
CREATE USER alice;
CREATE USER bob;

CREATE TABLEGROUP alice_grp OWNER alice;
CREATE TABLEGROUP alice_grp_2 OWNER alice;
CREATE TABLEGROUP alice_grp_3 OWNER alice;
CREATE TABLE bob_table (a INT) TABLEGROUP alice_grp;
CREATE TABLE bob_table_2 (a INT) TABLEGROUP alice_grp_2;
CREATE TABLE bob_table_3 (a INT) TABLEGROUP alice_grp_3;
CREATE TABLE alice_table (a INT);
ALTER TABLE bob_table OWNER TO bob;
ALTER TABLE bob_table_2 OWNER TO bob;
ALTER TABLE bob_table_3 OWNER TO bob;
ALTER TABLE alice_table OWNER TO alice;
-- Fails
\set VERBOSITY terse \\ -- suppress dependency details.
DROP TABLEGROUP alice_grp;  -- because bob_table depends on alice_grp
DROP USER alice;            -- because alice still owns tablegroups
DROP OWNED BY alice;        -- because bob's tables depend on alice's tablegroups
\set VERBOSITY default
-- Succeeds
DROP TABLEGROUP alice_grp_3 CASCADE;
DROP TABLE bob_table;

SELECT relname FROM pg_class WHERE relname LIKE 'bob%';

DROP TABLEGROUP alice_grp;    -- bob_table is gone, so alice_grp has no deps
DROP OWNED BY alice CASCADE;  -- CASCADE means we are allowed to drop bob's tables. We'll notify about any entities that she doesn't own.
DROP USER alice;              -- we dropped all of alice's entities, so we can remove her

SELECT relname FROM pg_class WHERE relname LIKE 'bob%';

--
-- Specifying tablegroup name for CREATE INDEX. These all fail.
--
CREATE INDEX ON tgroup_test1(col1) WITH (tablegroup=123);
CREATE INDEX ON tgroup_test1(col1) WITH (tablegroup=123, colocated=true);
CREATE INDEX ON tgroup_test1(col1) WITH (tablegroup=123) TABLEGROUP tgroup1;

--
-- Usage of SPLIT clause with TABLEGROUP should fail
--
CREATE TABLE tgroup_split (col1 int PRIMARY KEY) SPLIT INTO 3 TABLETS TABLEGROUP tgroup1;
CREATE TABLE tgroup_split (col1 int, col2 text) SPLIT INTO 3 TABLETS TABLEGROUP tgroup1;
CREATE INDEX ON tgroup_test1(col1) SPLIT AT VALUES((10), (20), (30));
CREATE INDEX ON tgroup_test1(col1) SPLIT AT VALUES((10), (20), (30)) TABLEGROUP tgroup2;
CREATE INDEX ON tgroup_test1(col1) SPLIT AT VALUES((10), (20), (30)) NO TABLEGROUP; -- should succeed
--
-- Test describes
--
CREATE TABLE tgroup_test4 (col1 int, col2 int) TABLEGROUP tgroup2;
CREATE INDEX ON tgroup_test4(col1);
CREATE INDEX ON tgroup_test4(col2);
-- Add comments
COMMENT ON TABLEGROUP tgroup1 IS 'Comment for Tablegroup 1';
COMMENT ON TABLEGROUP tgroup2 IS 'Comment for Tablegroup 2';
\dgr
\dgr+
\dgrt
\dgrt+
COMMENT ON TABLEGROUP tgroup2 IS NULL;
\dgr+ tgroup2
\dgrt tgroup2

-- Describe table
\d tgroup_test2
\d tgroup_test4
\d tgroup_test4_col1_idx
CREATE TABLEGROUP tgroup_describe1;
CREATE TABLEGROUP tgroup_describe2;
CREATE TABLE tgroup_describe (col1 int) TABLEGROUP tgroup_describe1;
CREATE INDEX ON tgroup_describe(col1);
CREATE INDEX ON tgroup_describe(col1) NO TABLEGROUP;
CREATE INDEX ON tgroup_describe(col1) TABLEGROUP tgroup_describe2;
\d tgroup_describe
--
-- DROP TABLEGROUP
--

DROP TABLEGROUP tgroup3;
-- These should fail. CREATE TABLE is to check that the row entry was deleted from pg_yb_tablegroup.
CREATE TABLE tgroup_test5 (col1 int, col2 int) TABLEGROUP tgroup3;
\set VERBOSITY terse \\ -- suppress dependency details.
DROP TABLEGROUP tgroup1;
DROP TABLEGROUP IF EXISTS tgroup1;
\set VERBOSITY default
DROP TABLEGROUP bad_tgroupname;
-- This drop should work now.
DROP TABLE tgroup_test1;
DROP TABLE tgroup_test2;
DROP INDEX tgroup_test3_col1_idx1;
DROP TABLE tgroup_with1;
DROP TABLEGROUP tgroup1;
DROP TABLEGROUP IF EXISTS tgroup1;
-- Create a tablegroup with the name of a dropped tablegroup.
CREATE TABLEGROUP tgroup1;
DROP TABLEGROUP IF EXISTS tgroup1;
CREATE TABLEGROUP tgroup1;

--
-- Assigning a tablespace to a tablegroup
--
CREATE TABLESPACE tblspc WITH (replica_placement='{"num_replicas": 1, "placement_blocks": [{"cloud":"cloud1","region":"datacenter1","zone":"rack1","min_num_replicas":1}]}');

-- These should fail
CREATE TABLEGROUP grp1 TABLESPACE nonexistentspc;
CREATE TABLEGROUP grp2 TABLESPACE pg_global;
-- These should succeeed
CREATE TABLEGROUP grp3 TABLESPACE tblspc;

SET default_tablespace = "tblspc";
CREATE TABLEGROUP grp4;
SET default_tablespace = '';

CREATE TABLE tgroup_test6 (col1 int, col2 int) TABLEGROUP grp3;
\dgr+
\dgrt+

--
-- Interactions with colocated database.
--

CREATE DATABASE db_colocated colocated=true;
\c db_colocated
-- This should fail.
CREATE TABLEGROUP tgroup1;
