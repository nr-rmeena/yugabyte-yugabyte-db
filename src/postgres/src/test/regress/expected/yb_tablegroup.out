--
-- YB_TABLEGROUP Testsuite: Testing Statments for TABLEGROUP.
--
\h CREATE TABLEGROUP
Command:     CREATE TABLEGROUP
Description: define a new tablegroup
Syntax:
CREATE TABLEGROUP tablegroup_name

\h ALTER TABLEGROUP
Command:     ALTER TABLEGROUP
Description: change the definition of a tablegroup
Syntax:
ALTER TABLEGROUP name RENAME TO new_name
ALTER TABLEGROUP name OWNER TO { new_owner | CURRENT_USER | SESSION_USER }

\h DROP TABLEGROUP
Command:     DROP TABLEGROUP
Description: remove a tablegroup
Syntax:
DROP TABLEGROUP name

--
-- pg_catalog alterations. Validate columns of pg_yb_tablegroup and oids.
--
\d pg_yb_tablegroup
            Table "pg_catalog.pg_yb_tablegroup"
    Column     |   Type    | Collation | Nullable | Default
---------------+-----------+-----------+----------+---------
 grpname       | name      |           | not null |
 grpowner      | oid       |           | not null |
 grptablespace | oid       |           | not null |
 grpacl        | aclitem[] |           |          |
 grpoptions    | text[]    |           |          |
Indexes:
    "pg_yb_tablegroup_oid_index" PRIMARY KEY, lsm (oid ASC)

SELECT oid, relname, reltype, relnatts FROM pg_class WHERE relname IN ('pg_yb_tablegroup', 'pg_yb_tablegroup_oid_index');
 oid  |          relname           | reltype | relnatts
------+----------------------------+---------+----------
 8036 | pg_yb_tablegroup           |    8038 |        5
 8037 | pg_yb_tablegroup_oid_index |       0 |        1
(2 rows)

SELECT oid, typname, typrelid FROM pg_type WHERE typname LIKE 'pg_yb_tablegroup';
 oid  |     typname      | typrelid
------+------------------+----------
 8038 | pg_yb_tablegroup |     8036
(1 row)

--
-- CREATE TABLEGROUP
--
CREATE TABLEGROUP tgroup1;
CREATE TABLEGROUP tgroup2;
CREATE TABLEGROUP tgroup3;
CREATE TABLE tgroup_test1 (col1 int, col2 int) TABLEGROUP tgroup1;
CREATE TABLE tgroup_test2 (col1 int, col2 int) TABLEGROUP tgroup1;
CREATE TABLE nogroup (col1 int) NO TABLEGROUP; -- fail
ERROR:  Cannot use NO TABLEGROUP in CREATE TABLE statement.
SELECT grpname FROM pg_yb_tablegroup;
 grpname
---------
 tgroup1
 tgroup2
 tgroup3
(3 rows)

SELECT relname
    FROM (SELECT relname, unnest(reloptions) AS opts FROM pg_class) s
    WHERE opts LIKE '%tablegroup%';
   relname
--------------
 tgroup_test1
 tgroup_test2
(2 rows)

CREATE INDEX ON tgroup_test1(col2);
CREATE TABLE tgroup_test3 (col1 int, col2 int) TABLEGROUP tgroup2;
-- Index opt out - should not show up in following SELECT
CREATE INDEX ON tgroup_test3(col1) NO TABLEGROUP;
-- Index explicitly specify tablegroup other than that of indexed table
CREATE INDEX ON tgroup_test3(col1) TABLEGROUP tgroup1;
SELECT s.relname, pg_yb_tablegroup.grpname
    FROM (SELECT relname, unnest(reloptions) AS opts FROM pg_class) s, pg_yb_tablegroup
    WHERE opts LIKE CONCAT('%tablegroup=', CAST(pg_yb_tablegroup.oid AS text), '%');
        relname         | grpname
------------------------+---------
 tgroup_test1           | tgroup1
 tgroup_test2           | tgroup1
 tgroup_test1_col2_idx  | tgroup1
 tgroup_test3           | tgroup2
 tgroup_test3_col1_idx1 | tgroup1
(5 rows)

-- These should fail.
CREATE TABLEGROUP tgroup1;
ERROR:  tablegroup "tgroup1" already exists
CREATE TABLE tgroup_test (col1 int, col2 int) TABLEGROUP bad_tgroupname;
ERROR:  tablegroup "bad_tgroupname" does not exist
CREATE TABLE tgroup_optout (col1 int, col2 int) WITH (colocated=false) TABLEGROUP tgroup1;
ERROR:  cannot use 'colocated=true/false' with tablegroup
CREATE TABLE tgroup_optout (col1 int, col2 int) WITH (colocated=true) TABLEGROUP tgroup1;
ERROR:  cannot use 'colocated=true/false' with tablegroup
CREATE TABLE tgroup_optout (col1 int, col2 int) WITH (colocated=false) TABLEGROUP bad_tgroupname;
ERROR:  tablegroup "bad_tgroupname" does not exist
CREATE TEMP TABLE tgroup_temp (col1 int, col2 int) TABLEGROUP tgroup1;
ERROR:  Cannot use TABLEGROUP with TEMP table.
-- Can use WITH to create a tablegroup
CREATE TABLE tgroup_with1 (col1 int, col2 int) WITH (tablegroup=16385);
-- Cannot use tablegroups and colocated=true/false
CREATE TABLE tgroup_with2 (col1 int, col2 int) WITH (tablegroup=16385, colocated=true);
ERROR:  cannot use 'colocated=true/false' with tablegroup
CREATE TABLE tgroup_with2 (col1 int, col2 int) WITH (tablegroup=16385, colocated=false);
ERROR:  cannot use 'colocated=true/false' with tablegroup
-- Cannot specify tablegroup OID and tablegroup name
CREATE TABLE tgroup_with3 (col1 int, col2 int) WITH (tablegroup=16385) TABLEGROUP tgroup1;
ERROR:  cannot specify both tablegroup and tablegroup oid
-- Cannot use an invalid tablegroup OID
CREATE TABLE tgroup_with4 (col1 int, col2 int) WITH (tablegroup=123);
ERROR:  tablegroup with oid 123 does not exist
--
-- Specifying tablegroup name for CREATE INDEX. These all fail.
--
CREATE INDEX ON tgroup_test1(col1) WITH (tablegroup=123);
ERROR:  cannot supply tablegroup through WITH clause
CREATE INDEX ON tgroup_test1(col1) WITH (tablegroup=123, colocated=true);
ERROR:  cannot supply tablegroup through WITH clause
CREATE INDEX ON tgroup_test1(col1) WITH (tablegroup=123) TABLEGROUP tgroup1;
ERROR:  cannot supply tablegroup through WITH clause
--
-- Usage of SPLIT clause with TABLEGROUP should fail
--
CREATE TABLE tgroup_split (col1 int PRIMARY KEY) SPLIT INTO 3 TABLETS TABLEGROUP tgroup1;
ERROR:  Cannot use TABLEGROUP with SPLIT.
CREATE TABLE tgroup_split (col1 int, col2 text) SPLIT INTO 3 TABLETS TABLEGROUP tgroup1;
ERROR:  Cannot use TABLEGROUP with SPLIT.
CREATE INDEX ON tgroup_test1(col1) SPLIT AT VALUES((10), (20), (30));
ERROR:  Cannot use TABLEGROUP with SPLIT.
DETAIL:  Please supply NO TABLEGROUP to opt-out of indexed table's tablegroup.
CREATE INDEX ON tgroup_test1(col1) SPLIT AT VALUES((10), (20), (30)) TABLEGROUP tgroup2;
ERROR:  Cannot use TABLEGROUP with SPLIT.
CREATE INDEX ON tgroup_test1(col1) SPLIT AT VALUES((10), (20), (30)) NO TABLEGROUP; -- should succeed
--
-- Test describes
--
CREATE TABLE tgroup_test4 (col1 int, col2 int) TABLEGROUP tgroup2;
CREATE INDEX ON tgroup_test4(col1);
CREATE INDEX ON tgroup_test4(col2);
-- Add comments
COMMENT ON TABLEGROUP tgroup1 IS 'Comment for Tablegroup 1';
COMMENT ON TABLEGROUP tgroup2 IS 'Comment for Tablegroup 2';
\dgr
List of tablegroups
  Name   |  Owner
---------+----------
 tgroup1 | yugabyte
 tgroup2 | yugabyte
 tgroup3 | yugabyte
(3 rows)

\dgr+
                                   List of tablegroups
  Name   |  Owner   | Access privileges |       Description        | Tablespace | Options
---------+----------+-------------------+--------------------------+------------+---------
 tgroup1 | yugabyte |                   | Comment for Tablegroup 1 |            |
 tgroup2 | yugabyte |                   | Comment for Tablegroup 2 |            |
 tgroup3 | yugabyte |                   |                          |            |
(3 rows)

\dgrt
                      List of tablegroup tables
 Group Name | Group Owner |          Name          | Type  |  Owner
------------+-------------+------------------------+-------+----------
 tgroup1    | yugabyte    | tgroup_test1           | table | yugabyte
 tgroup1    | yugabyte    | tgroup_test2           | table | yugabyte
 tgroup1    | yugabyte    | tgroup_test1_col2_idx  | index | yugabyte
 tgroup1    | yugabyte    | tgroup_test3_col1_idx1 | index | yugabyte
 tgroup1    | yugabyte    | tgroup_with1           | table | yugabyte
 tgroup2    | yugabyte    | tgroup_test4_col2_idx  | index | yugabyte
 tgroup2    | yugabyte    | tgroup_test4           | table | yugabyte
 tgroup2    | yugabyte    | tgroup_test3           | table | yugabyte
 tgroup2    | yugabyte    | tgroup_test4_col1_idx  | index | yugabyte
(9 rows)

\dgrt+
                                                                             List of tablegroup tables
 Group Name | Group Owner | Access privileges |    Group Description     | Group Tablespace | Group Options |          Name          | Type  |  Owner   | Rel Description |  Size
------------+-------------+-------------------+--------------------------+------------------+---------------+------------------------+-------+----------+-----------------+---------
 tgroup1    | yugabyte    |                   | Comment for Tablegroup 1 |                  |               | tgroup_test1           | table | yugabyte |                 | 0 bytes
 tgroup1    | yugabyte    |                   | Comment for Tablegroup 1 |                  |               | tgroup_test2           | table | yugabyte |                 | 0 bytes
 tgroup1    | yugabyte    |                   | Comment for Tablegroup 1 |                  |               | tgroup_test1_col2_idx  | index | yugabyte |                 | 0 bytes
 tgroup1    | yugabyte    |                   | Comment for Tablegroup 1 |                  |               | tgroup_test3_col1_idx1 | index | yugabyte |                 | 0 bytes
 tgroup1    | yugabyte    |                   | Comment for Tablegroup 1 |                  |               | tgroup_with1           | table | yugabyte |                 | 0 bytes
 tgroup2    | yugabyte    |                   | Comment for Tablegroup 2 |                  |               | tgroup_test4_col2_idx  | index | yugabyte |                 | 0 bytes
 tgroup2    | yugabyte    |                   | Comment for Tablegroup 2 |                  |               | tgroup_test4           | table | yugabyte |                 | 0 bytes
 tgroup2    | yugabyte    |                   | Comment for Tablegroup 2 |                  |               | tgroup_test3           | table | yugabyte |                 | 0 bytes
 tgroup2    | yugabyte    |                   | Comment for Tablegroup 2 |                  |               | tgroup_test4_col1_idx  | index | yugabyte |                 | 0 bytes
(9 rows)

COMMENT ON TABLEGROUP tgroup2 IS NULL;
\dgr+ tgroup2
                             List of tablegroups
  Name   |  Owner   | Access privileges | Description | Tablespace | Options
---------+----------+-------------------+-------------+------------+---------
 tgroup2 | yugabyte |                   |             |            |
(1 row)

\dgrt tgroup2
                      List of tablegroup tables
 Group Name | Group Owner |         Name          | Type  |  Owner
------------+-------------+-----------------------+-------+----------
 tgroup2    | yugabyte    | tgroup_test3          | table | yugabyte
 tgroup2    | yugabyte    | tgroup_test4          | table | yugabyte
 tgroup2    | yugabyte    | tgroup_test4_col1_idx | index | yugabyte
 tgroup2    | yugabyte    | tgroup_test4_col2_idx | index | yugabyte
(4 rows)

-- Describe table
\d tgroup_test2
            Table "public.tgroup_test2"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 col1   | integer |           |          |
 col2   | integer |           |          |
Tablegroup: "tgroup1"

\d tgroup_test4
            Table "public.tgroup_test4"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 col1   | integer |           |          |
 col2   | integer |           |          |
Indexes:
    "tgroup_test4_col1_idx" lsm (col1 ASC), tablegroup "tgroup2"
    "tgroup_test4_col2_idx" lsm (col2 ASC), tablegroup "tgroup2"
Tablegroup: "tgroup2"

\d tgroup_test4_col1_idx
 Index "public.tgroup_test4_col1_idx"
 Column |  Type   | Key? | Definition
--------+---------+------+------------
 col1   | integer | yes  | col1
lsm, for table "public.tgroup_test4"
Tablegroup: "tgroup2"

CREATE TABLEGROUP tgroup_describe1;
CREATE TABLEGROUP tgroup_describe2;
CREATE TABLE tgroup_describe (col1 int) TABLEGROUP tgroup_describe1;
CREATE INDEX ON tgroup_describe(col1);
CREATE INDEX ON tgroup_describe(col1) NO TABLEGROUP;
CREATE INDEX ON tgroup_describe(col1) TABLEGROUP tgroup_describe2;
\d tgroup_describe
          Table "public.tgroup_describe"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 col1   | integer |           |          |
Indexes:
    "tgroup_describe_col1_idx" lsm (col1 ASC), tablegroup "tgroup_describe1"
    "tgroup_describe_col1_idx1" lsm (col1 ASC)
    "tgroup_describe_col1_idx2" lsm (col1 ASC), tablegroup "tgroup_describe2"
Tablegroup: "tgroup_describe1"

--
-- DROP TABLEGROUP
--
DROP TABLEGROUP tgroup3;
-- These should fail. CREATE TABLE is to check that the row entry was deleted from pg_yb_tablegroup.
CREATE TABLE tgroup_test5 (col1 int, col2 int) TABLEGROUP tgroup3;
ERROR:  tablegroup "tgroup3" does not exist
DROP TABLEGROUP tgroup1;
ERROR:  tablegroup "tgroup1" is not empty
DROP TABLEGROUP bad_tgroupname;
ERROR:  tablegroup "bad_tgroupname" does not exist
-- This drop should work now.
DROP TABLE tgroup_test1;
DROP TABLE tgroup_test2;
DROP INDEX tgroup_test3_col1_idx1;
DROP TABLE tgroup_with1;
DROP TABLEGROUP tgroup1;
-- Create a tablegroup with the name of a dropped tablegroup.
CREATE TABLEGROUP tgroup1;
--
-- Assigning a tablespace to a tablegroup
--
CREATE TABLESPACE tblspc WITH (replica_placement='{"num_replicas": 1, "placement_blocks": [{"cloud":"cloud1","region":"datacenter1","zone":"rack1","min_num_replicas":1}]}');
-- These should fail
CREATE TABLEGROUP grp1 TABLESPACE nonexistentspc;
ERROR:  tablespace "nonexistentspc" does not exist
CREATE TABLEGROUP grp2 TABLESPACE pg_global;
ERROR:  tablegroups cannot be placed in pg_global tablespace
-- These should succeeed
CREATE TABLEGROUP grp3 TABLESPACE tblspc;
SET default_tablespace = "tblspc";
CREATE TABLEGROUP grp4;
SET default_tablespace = '';
CREATE TABLE tgroup_test6 (col1 int, col2 int) TABLEGROUP grp3;
\dgr+
                                 List of tablegroups
       Name       |  Owner   | Access privileges | Description | Tablespace | Options
------------------+----------+-------------------+-------------+------------+---------
 grp3             | yugabyte |                   |             | tblspc     |
 grp4             | yugabyte |                   |             | tblspc     |
 tgroup1          | yugabyte |                   |             |            |
 tgroup2          | yugabyte |                   |             |            |
 tgroup_describe1 | yugabyte |                   |             |            |
 tgroup_describe2 | yugabyte |                   |             |            |
(6 rows)

\dgrt+
                                                                              List of tablegroup tables
    Group Name    | Group Owner | Access privileges | Group Description | Group Tablespace | Group Options |           Name            | Type  |  Owner   | Rel Description |  Size
------------------+-------------+-------------------+-------------------+------------------+---------------+---------------------------+-------+----------+-----------------+---------
 grp3             | yugabyte    |                   |                   | tblspc           |               | tgroup_test6              | table | yugabyte |                 | 0 bytes
 tgroup2          | yugabyte    |                   |                   |                  |               | tgroup_test4              | table | yugabyte |                 | 0 bytes
 tgroup2          | yugabyte    |                   |                   |                  |               | tgroup_test4_col1_idx     | index | yugabyte |                 | 0 bytes
 tgroup2          | yugabyte    |                   |                   |                  |               | tgroup_test4_col2_idx     | index | yugabyte |                 | 0 bytes
 tgroup2          | yugabyte    |                   |                   |                  |               | tgroup_test3              | table | yugabyte |                 | 0 bytes
 tgroup_describe1 | yugabyte    |                   |                   |                  |               | tgroup_describe_col1_idx  | index | yugabyte |                 | 0 bytes
 tgroup_describe1 | yugabyte    |                   |                   |                  |               | tgroup_describe           | table | yugabyte |                 | 0 bytes
 tgroup_describe2 | yugabyte    |                   |                   |                  |               | tgroup_describe_col1_idx2 | index | yugabyte |                 | 0 bytes
(8 rows)

--
-- Interactions with colocated database.
--
CREATE DATABASE db_colocated colocated=true;
\c db_colocated
-- This should fail.
CREATE TABLEGROUP tgroup1;
ERROR:  cannot use tablegroups in a colocated database
